analysis/R/decode_test.R

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/Rscript
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+library(RUnit)
+library(abind)
+
+source('analysis/R/decode.R')
+source('tests/gen_counts.R')
+
+L1Distance <- function(X, Y) {
+  # Computes the L1 distance between two named vectors
+  common <- intersect(names(X), names(Y))
+  union <- rbind(X[common], Y[common])
+
+  (sum(abs(union[1,]-union[2,])) + sum(X[!names(X) %in% common])
+                                 + sum(Y[!names(Y) %in% common])) / 2
+}
+
+LInfDistance <- function(X, Y) {
+  # Computes the L1 distance between two named vectors
+  common <- intersect(names(X), names(Y))
+  union <- rbind(X[common], Y[common])
+
+  max(abs(union[1,]-union[2,]),
+      X[!names(X) %in% common],
+      Y[!names(Y) %in% common])
+}
+
+MatrixVectorMerge <- function(mat, vec) {
+  # Attaches a vector to a matrix, matching corresponding column names
+
+  mat_only <- setdiff(colnames(mat), names(vec))
+  vec_only <- setdiff(names(vec), colnames(mat))
+
+  # extend the vector with missing columns
+  vec_long <- c(vec, setNames(rep(NA, length(mat_only)), mat_only))
+
+  # extend the matrix with missing columns
+  newcols <- matrix(NA, nrow = nrow(mat), ncol = length(vec_only))
+  colnames(newcols) <- vec_only
+  mat_long <- cbind(mat, newcols)
+
+  # Now vec and mat have the same columns, but in the wrong order. Sort the
+  # columns lexicographically.
+  if(length(vec_long) > 0) {
+    mat_long <- mat_long[, order(colnames(mat_long)), drop = FALSE]
+    vec_long <- vec_long[order(names(vec_long))]
+  }
+
+  rbind(mat_long, vec_long)
+}
+
+RunMultipleTests <- function(title, fun, repetitions, ...) {
+  # Run a function with an annotated progress indicator
+  cat(title, ": ")
+
+  if(repetitions == 1) {
+    # only run once
+    fun(...)
+
+    cat(" Done.")
+  }
+  else {  # run multiple times
+    pb <- txtProgressBar(min = 0, max = repetitions,
+                         width = getOption("width") - 20 - nchar(title))
+
+    for(i in 1:repetitions) {
+      setTxtProgressBar(pb, i)
+      fun(...)
+    }
+    cat(" Done.")
+    close(pb)
+  }
+}
+
+TestEstimatesAndStdsHelper <- function(params, map, pdf, total) {
+  # Helper function for TestEstimateBloomCounts.
+  partition <- RandomPartition(total, pdf)
+  counts <- GenerateCounts(params, map, partition, 1)
+  e <- EstimateBloomCounts(params, counts)
+
+  results$estimates <<- abind(results$estimates, e$estimates, along = 3)
```

Comment:
Ditto, I think the use of <<- is too subtle -- how about just passing in a list() of everything you need as the first param?